### PR TITLE
fix(tron): fee estimator uses /wallet/ not /walletsolidity/ endpoint

### PR DESCRIPTION
**Pull request by gomesalexandre**
Tackles BUG-5 (LOW) from the 2026-05-25 Tron audit.

## what

`getTrc20TransferFee` was calling `walletsolidity/triggerconstantcontract`. The `walletsolidity` endpoints lag up to 20 confirmations behind the canonical node state - this causes spurious "account not found" errors for freshly-created contract addresses (TRC-20 tokens on new accounts).

The rest of the Tron stack already uses the non-solidified `/wallet/` endpoints:
- `getTronBlockRefs` - `wallet/getblock`
- `broadcastTronTx` - `wallet/broadcasthex`
- `getTronAccount` - `wallet/getaccount`

One-liner: `walletsolidity/triggerconstantcontract` -> `wallet/triggerconstantcontract`.

## grep receipt - no other walletsolidity callers in sdk

```
$ grep -rn "walletsolidity" packages/ --include="*.ts" --include="*.js" | grep -v ".test."
(no output)
```

Zero production callers after this fix.

## jest receipt

```
 RUN  v4.1.6 /private/tmp/sdk-tron-bug5

 ✓ .../resolvers/tron/fee.test.ts > getTrc20TransferFee > calls /wallet/triggerconstantcontract, not /walletsolidity/... 3ms
 ✓ .../resolvers/tron/fee.test.ts > getTrc20TransferFee > returns (energy_used + energy_penalty) * 280n as fee 0ms
 ✓ .../resolvers/tron/fee.test.ts > getTrc20TransferFee > defaults missing energy fields to 0 0ms

 Test Files  1 passed (1)
      Tests  3 passed (3)
```

Full core suite: 71 test files, 625 tests, all green.

🤖 Agent-generated

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Tron TRC-20 transfer fee calculation to reliably handle edge cases and prevent invalid fee estimates.

* **Tests**
  * Added comprehensive test coverage for Tron fee resolution.

[Review Change Stack](https://app.coderabbit.ai/change-stack/vultisig/vultisig-sdk/pull/546?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by NeOMakinG — COMMENTED** [review]
gm — multi-lane pass: . conceptual-only (no runtime this pass). comment-only.

## verdict
COMMENTED

🤖 reviewed via [Claude Code](https://claude.com/claude-code) with codex `gpt-5.3-codex` conceptual pass.

File: packages/core/mpc/keysign/chainSpecific/resolvers/tron/fee.test.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  it('documents negative energy_used / energy_penalty behavior (no clamping in current impl)', async () => {
+    // TronGrid edge cases can return negative values; current code does not clamp,
+    // so the returned fee is negative. This test pins the existing behavior so any
+    // future change (e.g. adding Math.max(0, ...) clamping) is a conscious decision.
+    mockQueryUrl.mockResolvedValue({ energy_used: -5000, energy_penalty: -1000 })
+
+    const fee = await getTrc20TransferFee({ coin, amount: 1_000_000n, receiver: 'TJDENsfBJs4RFETt1X1W8wMDc8M5XnJhd' })
+
+    // (-5000 + -1000) * 280 = -1_680_000n — negative because no clamp exists yet
+    expect(fee).toBe(-1_680_000n)
```
Comment (NeOMakinG) [inline packages/core/mpc/keysign/chainSpecific/resolvers/tron/fee.test.ts]: `should-fix:` the test documents negative energy returning a negative fee (`-1_680_000n`) but does not note the downstream consequence: this value flows as `gasEstimation` into `feeLimit` via `Long.fromString(gasEstimation.toString())` in `signingInputs/resolvers/tron.ts` across all TRC-20 paths. `Long.fromString('-1680000')` does not throw — it encodes a negative int64 in the protobuf, which TronGrid rejects at broadcast. The guard in `sdk/src/chains/tron/tx.ts:391` is on the send-service path only; the MPC keysign path has no equivalent guard.

fix: add a clamp in `fee.ts` — `const totalSun = totalEnergy <= 0n ? 0n : totalEnergy * 280n` — and update this test to expect `0n` instead of `-1_680_000n`.

**Review by NeOMakinG — COMMENTED** [review]
triple-lane review (Claude correctness + GLM chain-specific + Codex path-trace). 3 tests: URL pin, arithmetic, error propagation. 2 edge-case pins (missing-field, negative energy). All 3 lanes agree: the one-liner fix is correct and the grep receipt in the PR body is accurate — zero remaining `walletsolidity` callers in non-test source. One `should-fix` surfaced by the Codex path-trace lane; no blockers.

## testing approach

Code-only triple-lane review. Emulator unavailable. Runtime receipt deferred — no behavioral change, just endpoint path.

## should-fix

1. **`fee.ts` negative gasEstimation flows unguarded into keysign `feeLimit`** — `getTrc20TransferFee` can return a negative `bigint` when TronGrid returns `energy_used < 0` (documented in the new negative-energy test). The result flows as `tronSpecific.gasEstimation` directly into `feeLimit` via `Long.fromString(gasEstimation.toString())` in `signingInputs/resolvers/tron.ts:57,86,130,141,213,276`. `Long.fromString('-1680000')` does not throw — it encodes a negative int64 in the protobuf, which TronGrid rejects at broadcast. The guard in `sdk/src/chains/tron/tx.ts:391` (`feeLimit <= 0n` throws) is on the send-service path only; the MPC keysign path has no equivalent.

   action: add `const totalSun = totalEnergy <= 0n ? 0n : totalEnergy * 280n` at `fee.ts:56`, OR add the guard in `getTronChainSpecific` after calling `getTrc20TransferFee`. Zero is safe — a zero feeLimit fails at broadcast with a clear error rather than a confusing negative-field rejection.

## suggestion

1. **`TriggerContractResponse` misses `energy_required`** — `/wallet/triggerconstantcontract` can return `energy_required` (server-computed total = used + penalty). The SDK's own `rpc.ts:329` handles this: `res.energy_required ?? res.energy_used ?? 0`. The fee.ts arithmetic of `energy_used + energy_penalty` is equivalent when both fields are present, but if TronGrid omits individual fields and sets `energy_required` instead, fee.ts silently returns a lower estimate.

   fix: add `energy_required?: number` to `TriggerContractResponse` and use `responseData.energy_required ?? (energyUsed + energyPenalty)`.

## q

1. The `280n` SUN/energy multiplier is hardcoded. iOS `TronService.swift` fetches `energy_price` dynamically from `/wallet/getchainparameters` (currently 420 SUN on mainnet). This PR doesn't introduce the discrepancy, but is there a tracked issue for aligning the TS energy-price source with the iOS implementation?

## pre-existing bugs i bumped into

- `280n` hardcoded energy price vs iOS 420 SUN dynamic fetch — pre-existing, not introduced here.
- `TriggerContractResponse` missing `Error?/error?` bare-envelope guard (all other Tron helpers in the SDK have it) — pre-existing.

## risk

Merging as-is is safe for the normal path (positive energy simulation). The unguarded negative-gasEstimation path exists pre-PR; production TronGrid should never return negative energy on a valid contract call.

## verdict

COMMENTED — one `should-fix` (negative feeLimit guard in keysign path). Core one-liner fix and test coverage are correct.

🤖 reviewed via [Claude Code](https://claude.com/claude-code) with codex `gpt-5.4` adversarial path-trace + GLM `glm-4.6` chain-edge lane. macOS.

**Review by NeOMakinG — APPROVED** [review]
**fix landed** — negative-energy clamp + `energy_required` preference

Changes pushed in commit `2774160d`:

**`fee.ts`**
- `energy_required` added to `TriggerContractResponse` type; when present it is used as the pre-summed total (bypasses `energy_used + energy_penalty` to avoid double-counting)
- `totalEnergy <= 0n ? 0n : totalEnergy * 280n` — clamp guards the keysign path; negative totals no longer reach `Long.fromString` in `signingInputs/resolvers/tron.ts`

**`fee.test.ts`**
- Negative-energy test updated: expectation changed from `-1_680_000n` to `0n`, description updated to reflect clamping
- New test: `prefers energy_required over energy_used + energy_penalty when present` — verifies `40000 * 280 = 11_200_000n` (not `35000 * 280`)

All 628 core tests pass locally (`yarn test:core`).

File: packages/core/mpc/keysign/chainSpecific/resolvers/tron/fee.test.ts (hunk 1 above)
Comment (gomesalexandre) [reply packages/core/mpc/keysign/chainSpecific/resolvers/tron/fee.test.ts]: @NeOMakinG tackled in [c3cb26ca](https://github.com/vultisig/vultisig-sdk/commit/c3cb26ca). Added the clamp in `fee.ts` (`if (totalEnergy <= 0n) return 0n`) - mirrors the existing send-service guard at `sdk/src/chains/tron/tx.ts:391`. Test updated to assert 0n instead of -1_680_000n. All 5 fee tests pass. tyvm 🙇
